#! /bin/sh /usr/share/dpatch/dpatch-run
## 08_support-include-statement.dpatch by  <jcorbier@ubuntu.com>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: Upstream patch which enables include statements in config files.

@DPATCH@

--- orig/cfparse.y	2006/04/11 04:33:52	1.3
+++ dest/cfparse.y	2006/05/05 05:39:51	1.5
@@ -91,6 +91,7 @@
 long long cf_refreshtime = -1;
 
 extern int yylex __P((void));
+extern int cfswitch_buffer __P((char *));
 static int add_namelist __P((struct cf_namelist *, struct cf_namelist **));
 static void cleanup __P((void));
 static void cleanup_namelist __P((struct cf_namelist *));
@@ -111,6 +112,7 @@
 %token AUTHNAME RDM KEY
 %token KEYINFO REALM KEYID SECRET KEYNAME EXPIRE
 %token ADDRPOOL POOLNAME RANGE TO ADDRESS_POOL
+%token INCLUDE
 
 %token NUMBER SLASH EOS BCL ECL STRING QSTRING PREFIX INFINITY
 %token COMMA
@@ -151,6 +153,7 @@
 	|	authentication_statement
 	|	key_statement
 	|	addrpool_statement
+	|	include_statement
 	;
 
 interface_statement:
@@ -336,6 +339,17 @@
 	}
 	;
 
+include_statement:
+	INCLUDE QSTRING EOS
+	{
+		if (cfswitch_buffer($2)) {
+			free($2);
+			return (-1);
+		}
+		free($2);
+	}
+	;
+
 addrpool_statement:
 	ADDRPOOL POOLNAME BCL declarations ECL EOS
 	{
@@ -388,6 +402,7 @@
 
 		$$ = l;
 	}
+	;
 
 declarations:
 		{ $$ = NULL; }
@@ -520,6 +535,7 @@
 
 			$$ = $1;
 		}
+	;
 
 dhcpoption:
 		RAPID_COMMIT
@@ -693,6 +709,7 @@
 
 			$$ = pconf;
 		}
+	;
 
 prefixparam:
 		STRING SLASH NUMBER duration
@@ -752,6 +769,7 @@
 
 			$$ = pconf;
 		}
+	;
 
 poolparam:
 		STRING duration
--- orig/cftoken.l	2006/04/11 04:33:52	1.3
+++ dest/cftoken.l	2006/04/26 09:32:46	1.4
@@ -60,6 +60,18 @@
 char *configfilename;
 int lineno = 1;
 
+
+/* Recursion limit for includes */
+#define MAX_INCLUDE_DEPTH 10
+
+static struct include_stack {
+	char *path;
+	YY_BUFFER_STATE state;
+	int lineno;
+} incstack[MAX_INCLUDE_DEPTH];
+int incstackp = 0;
+
+
 static int yy_first_time = 1;
 static int yyerrorcount = 0;
  
@@ -111,6 +123,7 @@
 %s S_KEY
 %s S_SECRET
 %s S_ADDRPOOL
+%s S_INCL
 
 %%
 %{
@@ -267,6 +280,15 @@
 }
 <S_CNF>expire { DECHO; return (EXPIRE); }
 
+	/* include */
+<S_CNF>include { DECHO; BEGIN S_INCL; return (INCLUDE); }
+<S_INCL>{quotedstring} {
+	DECHO;
+	yylval.str = strdup(yytext);
+	BEGIN S_CNF;
+	return (QSTRING);
+}
+
 	/* quoted string */
 {quotedstring} {
 		DECHO;
@@ -296,6 +318,18 @@
 		return (STRING);
 	}
 
+<<EOF>>	{
+		if (--incstackp < 0)
+			yyterminate();
+		else {
+			yy_delete_buffer(YY_CURRENT_BUFFER);
+			free(incstack[incstackp + 1].path);
+			configfilename = incstack[incstackp].path;
+			lineno = incstack[incstackp].lineno;
+			yy_switch_to_buffer(incstack[incstackp].state);
+		}
+	}
+
 %%
 static void
 cfdebug_print(w, t, l)
@@ -353,6 +387,39 @@
 }
 
 int
+cfswitch_buffer(incl)
+	char *incl;
+{
+	char *path = qstrdup(incl);
+	FILE *fp;
+
+	if (incstackp >= MAX_INCLUDE_DEPTH) {
+		dprintf(LOG_ERR, FNAME, "cfparse: includes nested too deeply");
+		return (-1);
+	}
+	incstack[incstackp].path = configfilename;
+	incstack[incstackp].state = YY_CURRENT_BUFFER;
+	incstack[incstackp].lineno = lineno;
+
+	fp = fopen(path, "r");
+	if (fp == NULL) {
+		dprintf(LOG_ERR, FNAME, "cfparse: fopen(%s): %s",
+			path, strerror(errno));
+		if (errno == ENOENT)
+			return (0);
+		return (-1);
+	}
+	incstackp++;
+	configfilename = path;
+	lineno = 1;
+	yy_switch_to_buffer(yy_create_buffer(fp, YY_BUF_SIZE));
+
+	BEGIN(S_CNF);
+
+	return (0);
+}
+
+int
 cfparse(conf)
 	char *conf;
 {
--- orig/config.c	2006/04/11 04:33:52	1.3
+++ dest/config.c	2006/04/26 09:32:46	1.4
@@ -127,7 +127,6 @@
 static void clear_authinfo __P((struct authinfo *));
 static int configure_duid __P((char *, struct duid *));
 static int get_default_ifid __P((struct prefix_ifconf *));
-static char *qstrdup __P((char *));
 static void clear_poolconf __P((struct pool_conf *));
 static struct pool_conf *create_pool __P((char *, struct dhcp6_range *));
 struct host_conf *find_dynamic_hostconf __P((struct duid *));
@@ -1867,7 +1866,7 @@
 	return (NULL);
 }
 
-static char *
+char *
 qstrdup(qstr)
 	char *qstr;
 {
--- orig/config.h	2006/04/11 04:33:52	1.4
+++ dest/config.h	2006/04/26 09:32:46	1.5
@@ -318,3 +318,4 @@
 	struct in6_addr *));
 struct host_conf *create_dynamic_hostconf __P((struct duid *,
 	struct dhcp6_poolspec *));
+extern char *qstrdup __P((char *));
--- orig/dhcp6s.conf.5	2006/01/26 07:15:56	1.4
+++ dest/dhcp6s.conf.5	2006/04/26 09:32:46	1.5
@@ -60,6 +60,18 @@
 and
 .Ar gif1.
 .\"
+.Sh Include statement
+An include statement specifies another configuration file to be included.
+The format of an include statement is as follows:
+.Bl -tag -width Ds -compact
+.It Xo
+.Ic include Ar \(dqfilename\(dq ;
+.Xc
+Where
+.Ar \(dqfilename\(dq
+is the name (full path) of the file to be included.
+.El
+.\"
 .Sh Option statement
 An option statement specifies configuration parameters provided for
 every client.
